### PR TITLE
OCPBUGS-385: Capability to override default channel

### PR DESCRIPTION
**Pull request by lmzuccarelli**
# Description

This fix address the issues found  [here](https://issues.redhat.com/browse/OCPBUGS-385) .

The TL;DR is that dev teams can change the default channel from one release to the next, this causes customers who have set CICD pipelines to have to change their ImageSetConfig each time a version changes.
This fix allow the use of a field in the ImageSetConfig to override the default channel and specify any channel and use the HEAD (latest version of that channel). 

It means that the customer needs to be aware of the caveats not using the default channel when mirroring. This will mean we need to update our documents accordingly.  We will work with the docs team as to find the appropriate wording and track (link) the docs issue to this.

Fixes # [OCPBUGS-385](https://issues.redhat.com/browse/OCPBUGS-385)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

This has been tested locally and seems to working as expected.
We will need thorough testing from QE to ensure this fix works.

## Expected Outcome

The ImageSetConfig used is as follows

```kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v1alpha2
mirror:
  operators:
  - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.12
    packages:
    - name: elasticsearch-operator
      overrideDefaultChannel: true
      channels:
      - name: stable

``` 

**NB**  the overrideDefaultChannel flag is at the package level , also their can only be **one** channel specified.

Use the following cli to execute 

```
bin/oc-mirror --config test-isc.yml file://<some-path>
```
The head of the current channel "stable" is version "elasticsearch-operator.v5.8.0" 

To verify this  
- execute the command as shown above
- update the ImageSetConfig to use the defaultChannel (HEAD is the same elasticsearch-operator.v5.8.0)
- execute the oc command but ensure the contents are save in a different folder
- compare folders

### CODE REVIEW

**Review by sherine-k — COMMENTED** [review]
I like the simplicity of the solution. :+1:
TBH, I haven't tested this locally yet.

Some ImageSetConfigs used by customers contain several channels per package, so this fix might not help their use case. 
Nevertheless, this can come as a second step, once we've established that this solution is viable (ie the catalog is valid from the standpoint of OLM in a cluster).

**Comment by joeg-pro** [comment]
Its been noted on OCPBUGS-385 that this fix only works when exactly one channel is being mirrored, and if the ImageSetConfig specifies more than one then you get an error.  In my opinion, fixing the default-channel issue only for the singe-channel case is too restrictive.

As @sherine-k  suggested in an earlier comment,  instead of  ` overrideDefaultChannel ` being a boolean, maybe instead it should be a string (perhaps called `defaultChannel`)  which would let the user pick the channel they want as default.  This approach would be a tad more complicated for the user in the single-channel case (since you'd have to name the single channel rather than just specifying `true`) but has the big advantage that it would address the multi-channel case as well.

**Comment by lmzuccarelli** [comment]
UPDATE 

The ImageSetConfig used is as follows

apiVersion: mirror.openshift.io/v1alpha2
mirror:
  operators:
  - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.12
    packages:
    - name: elasticsearch-operator
      setDefaultChannel:  'stable'
      channels:
      - name:  'stable'
      - name:  'stable-5.6'

**NB**  the setDefaultChannel flag is at the package level and is now  a string value

Use the following cli to execute

bin/oc-mirror --config test-isc.yml file://<some-path>

**Comment by lmzuccarelli** [comment]
@zhouying7780 - when you get a chance could you PTAL and verify. I have made the suggested changes

**Review by sherine-k — COMMENTED** [review]
Looking great @lmzuccarelli!
I think this will help a lot of our customers.
Maybe add a small check to ensure that the setDefaultChannel is among the channels that are included in the imageSetConfig for that operator?

File: pkg/operator/diff/internal/diff.go [hunk 1]
```diff
@@ -185,22 +184,36 @@ func (g *DiffGenerator) Run(oldModel, newModel model.Model) (model.Model, error)
 	}
 
 	// Default channel may not have been copied, so set it to the new default channel here.
-	for _, outputPkg := range outputModel {
+	overrideSet := false
```
[the remark this answers is not in the document]
Comment (lmzuccarelli) [reply pkg/operator/diff/internal/diff.go:187]: Yes good catch - this loop is for every package and not channel

**Comment by lmzuccarelli** [comment]
> Looking great @lmzuccarelli! I think this will help a lot of our customers. Maybe add a small check to ensure that the setDefaultChannel is among the channels that are included in the imageSetConfig for that operator?

Yes good point - I have done that

File: pkg/operator/diff/internal/diff.go [hunk 2]
```diff
@@ -262,7 +275,7 @@ This can be resolved by one of the following changes:
 3) by changing the ImageSetConfiguration to filter channels or packages in such a way that it will include a package version that exists in the current default channel`)
 
 	// include a short message that does not mention any of the above to keep things simple
-	return fmt.Errorf("the current default channel %q for package %q could not be determined... ensure that your ImageSetConfiguration filtering criteria results in a package version that exists in the current default channel", newPackageDefaultChannelName, outputPkg.Name)
+	return fmt.Errorf("the current default channel %q for package %q could not be determined... ensure that your ImageSetConfiguration filtering criteria results in a package version that exists in the current default channel or use the overrideDefaultChannel: field", newPackageDefaultChannelName, outputPkg.Name)
```
[the remark this answers is not in the document]
Comment (lmzuccarelli) [reply pkg/operator/diff/internal/diff.go:278]: @sherine-k - I think the selectedDefaultChannel  sounds odd, any there was a comment by @joeg-pro about simple calling it defaultChanne; , which I'm in favor of, so if there are nor objections I'm going with that

[the remark this answers is not in the document]
Comment (sherine-k) [reply pkg/operator/diff/internal/diff.go:278]: perfect!

**Comment by lmzuccarelli** [comment]
@zhouying7780 - PTAL when you get a chance. See the comments above for verification

**Comment by lmzuccarelli** [comment]
@zhouying7780 - There is one extra validation that I should point out wrt ImageSetConfig

```
apiVersion: mirror.openshift.io/v1alpha2
mirror:
  operators:
  - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.12
    packages:
    - name: elasticsearch-operator
      defaultChannel:  'stable'
      channels:
       - name:  'stable-5.6'
```
In this scenario oc-mirror should error with a message "ensure to include 'stable' in the channels section

Here is the corrected ImageSetConfig 

```
apiVersion: mirror.openshift.io/v1alpha2
mirror:
  operators:
  - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.12
    packages:
    - name: elasticsearch-operator
      defaultChannel:  'stable'
      channels:
       - name: 'stable'
       - name:  'stable-5.6'
```

**Comment by zhouying7780** [comment]
@lmzuccarelli  Sure , I'm checking today.

**Review by sherine-k — APPROVED** [review]
/lgtm
